### PR TITLE
Add Context Parallel tutorial

### PR DESCRIPTION
**Pull request by XilunWu**
PyTorch 2.7 release of `torch.distributed.tensor.experimental._attention`

API: https://pytorch.org/docs/stable/distributed.tensor.html#torch.distributed.tensor.experimental.context_parallel
tutorial preview: https://docs-preview.pytorch.org/pytorch/tutorials/3319/prototype/context_parallel.html

### CODE REVIEW

File: prototype_source/context_parallel.rst [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        qkv = [
+            torch.rand(
+                (batch, nheads, qkv_len, dim),
+                dtype=dtype,
+                requires_grad=True,
+                device='cuda',
+            )
+            for _ in range(3)
+        ]
+        cp_qkv = [t.detach().clone() for t in qkv]
```
Comment (wconstab) [inline prototype_source/context_parallel.rst:151]: this line seems unintuitive. can we comment why we need this line?

File: prototype_source/context_parallel.rst [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+        with sdpa_kernel(backend):
+            with context_parallel(
```
Comment (wconstab) [inline prototype_source/context_parallel.rst:154]: we might want to add a section to the tutorial that introduces the 'context_parallel' API, explains what it does.  Or, we can explain it here in a code comment.  I think its doing at least 2 things that might not be obvious to people.
1) it shards the buffers, right?  I think this is the main thing to explain more.  Users reading this might not know if they should add another buffer to 'buffers' list.  For instance, in llama we had to also add 'freqs_cis' becuase it contains a seq dim.  Also, 'buffer_seq_dims' is not explained here.  

2) it hooks the CP version of the kernel up

File: prototype_source/context_parallel.rst [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                device='cuda',
+            )
+            for _ in range(3)
+        ]
+        # specify the SDPBackend to use
+        with sdpa_kernel(backend):
+            out = F.scaled_dot_product_attention(*qkv, is_causal=True)
+
+        # make a clean copy of QKV for output comparison
+        cp_qkv = [t.detach().clone() for t in qkv]
```
Comment (wconstab) [inline prototype_source/context_parallel.rst:156]: Ahh, so this is not even needed for cp just for the reference?

I wonder if it's better to delete the reference. That's more appropriate for a unit test, but for an example people usually want something that's minimal and copy able, and in this case they might be distracted by this line.

**Review by wconstab — APPROVED** [review]

File: prototype_source/context_parallel.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
+Introduction to Context Parallel
+======================================
+**Authors**: `Xilun Wu <https://github.com/XilunWu>`_, `Chien-Chin Huang <https://github.com/fegin>`__
+
+.. note::
+    |edit| View and edit this tutorial in `github <https://github.com/pytorch/tutorials/blob/main/prototype_source/context_parallel.rst>`__.
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:6]: standardize format as GitHub

File: prototype_source/context_parallel.rst [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+.. grid:: 2
+
+   .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
+      :class-card: card-prerequisites
+
+      * `Context Parallel APIs <https://pytorch.org/docs/stable/distributed.tensor.html#torch.distributed.tensor.experimental.context_parallel>`__
+      * `1M sequence training in torchtitan with Context Parallel <https://discuss.pytorch.org/t/distributed-w-torchtitan-breaking-barriers-training-long-context-llms-with-1m-sequence-length-in-pytorch-using-context-parallel/215082>`__
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:14]: standardize as TorchTitan

File: prototype_source/context_parallel.rst [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+Introduction
+------------
+
+Context Parallel is an approach used in large language model training to reduce peak activation size by sharding the long input sequence across multiple devices.
+It breaks the constraint on input sequence length resulting from peak memory usage on storing activations in Transformer blocks.
+
+The core of Context Parallel is Ring Attention, a novel parallel implementation of the Attention layer.
+Ring Attention shuffles the KV shards and calculates the partial attention scores,
+repeats until all KV shards have been used on each device.
+We implemented two Ring Attention variants: `pass-KV <https://arxiv.org/abs/2411.01783>`__ and `all-to-all <https://openreview.net/forum?id=WsRHpHH4s0>`__.
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:32]: I think it's best if docs avoid wording things as "we did X." I think we should stick to wording it likee "Two Ring Attention variants have been implemented" or "There are two Ring Attention variants."

File: prototype_source/context_parallel.rst [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        # specify the SDPBackend to use
+        with sdpa_kernel(backend):
+            out = F.scaled_dot_product_attention(*qkv, is_causal=True)
+
+        # make a clean copy of QKV for output comparison
+        cp_qkv = [t.detach().clone() for t in qkv]
+
+        with sdpa_kernel(backend):
+            # This `context_parallel()` performs two actions:
+            # 1. shard the tensor objects in `buffers` in-place along the dimension
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:160]: Capitalize these sentences in the comments.

File: prototype_source/context_parallel.rst [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            #    specified in `buffer_seq_dims`, the tensors in `buffers` and their
+            #    sharding dims in `buffer_seq_dims` are organized in the same order.
+            # 2. replace the execution of `F.scaled_dot_product_attention` with a
+            #    context-paralleled-enabled Ring Attention.
+            with context_parallel(
+                device_mesh, buffers=tuple(cp_qkv), buffer_seq_dims=(2, 2, 2)
+            ):
+                cp_out = F.scaled_dot_product_attention(*cp_qkv, is_causal=True)
+
+            # the output `cp_out` is still sharded in the same way as QKV
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:170]: Capitalize

File: prototype_source/context_parallel.rst [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        with context_parallel(
+            device_mesh, buffers=tuple(cp_qkv), buffer_seq_dims=(2, 2, 2)
+        ):
+            cp_out = F.scaled_dot_product_attention(*cp_qkv, is_causal=True)
+
+
+Conclusion
+----------
+
+In this tutorial, have learned how to parallelize the SDPA computation along the sequence dimension easily with our Context Parallel APIs. For
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:219]: Missing a word - "we have learned"

File: prototype_source/context_parallel.rst [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+Ring Attention shuffles the KV shards and calculates the partial attention scores,
+repeats until all KV shards have been used on each device.
+We implemented two Ring Attention variants: `pass-KV <https://arxiv.org/abs/2411.01783>`__ and `all-to-all <https://openreview.net/forum?id=WsRHpHH4s0>`__.
+The pass-KV approach all-gathers KV shards while performing the local SDPA (Scaled Dot Product Attention) then performs the rest when the communication completes.
+The all-to-all approach uses interleaved all-to-all collectives to ring shuffle KV shards to overlap the SDPA computation and the all-to-all communication
+necessary for the next SDPA.
+
+The Context Parallel APIs consist of two parts:
+
+1. ``context_parallel()`` allows users to create a Python context where the SDPA function (``torch.nn.functional.scaled_dot_product_attention``)
```
Comment (AlannaBurke) [inline prototype_source/context_parallel.rst:39]: Fix number formatting, it's not rendering correctly. 
Monosnap Introduction to Context Parallel — PyTorch Tutorials 2 7 0+cu126 documentation 2025-04-12 13-46-12

**Review by AlannaBurke — CHANGES_REQUESTED** [review]
Just some minor formatting fixes.

File: prototype_source/context_parallel.rst [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+------------
+
+Context Parallel is an approach used in large language model training to reduce peak activation size by sharding the long input sequence across multiple devices.
+It breaks the constraint on input sequence length resulting from peak memory usage on storing activations in Transformer blocks.
+
+The core of Context Parallel is Ring Attention, a novel parallel implementation of the Attention layer.
+Ring Attention shuffles the KV shards and calculates the partial attention scores,
+repeats until all KV shards have been used on each device.
+Two Ring Attention variants have been implemented: `pass-KV <https://arxiv.org/abs/2411.01783>`__ and `all-to-all <https://openreview.net/forum?id=WsRHpHH4s0>`__.
+The pass-KV approach all-gathers KV shards while performing the local SDPA (Scaled Dot Product Attention) then performs the rest when the communication completes.
```
Comment (fegin) [inline prototype_source/context_parallel.rst:33]: This is not correct. `pass-KV` means that KV will be passed from rank to rank. This can be implemented with allgather or all-to-all. So essentially, `all-to-all` is also `pass-kv`. The cited paper is also not correct as that paper uses `SendRecv` instead of `all-gather`.

We could say `all-gather based pass-KV` and `all-to-all based pass-KV` and cite llama3 paper for `all-gather based` and mentioned our implementation is slightly different because we also try to overlap the all-gather with SDPA.

File: prototype_source/context_parallel.rst [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+The all-to-all approach uses interleaved all-to-all collectives to ring shuffle KV shards to overlap the SDPA computation and the all-to-all communication
+necessary for the next SDPA.
+
+The Context Parallel APIs consist of two parts:
+
+1. ``context_parallel()`` allows users to create a Python context where the SDPA function (``torch.nn.functional.scaled_dot_product_attention``)
+   will be automatically replaced with Ring Attention. To shard Tensors along a dimension, simply pass the Tensors and their sharding dimensions to
+   argument ``buffers`` and ``buffer_seq_dims`` respectively.
```
Comment (fegin) [inline prototype_source/context_parallel.rst:41]: We should explain what tensors are required to be sharded. Specifically, we should mention that any tensors that will be used for sequence-dim related computation should be sharded, e.g., freq_cis.

**Review by fegin — APPROVED** [review]

**Review by fegin — CHANGES_REQUESTED** [review]
There is a wrong statement about `pass-kv`, we should change that.

File: prototype_source/context_parallel.rst [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+      * PyTorch 2.7 or later
+
+
+Introduction
+------------
+
+Context Parallel is an approach used in large language model training to reduce peak activation size by sharding the long input sequence across multiple devices.
+It breaks the constraint on input sequence length resulting from peak memory usage on storing activations in Transformer blocks.
+
+The core of Context Parallel is Ring Attention, a novel parallel implementation of the Attention layer.
```
Comment (fegin) [inline prototype_source/context_parallel.rst:29]: `Ring Attention` is used to optimize the performance of `Context Parallel`. So I prefer not to say the core of `Context Parallel` is `Ring Attention` as llama3 doesn't use it.

File: prototype_source/context_parallel.rst [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    overlap between the attention computation and the all-gather collective. For example, in the case of Llama3 training, we also shard ``freq_cis`` over the sequence dimension.
+2.  The all-to-all approach uses interleaved all-to-all collectives to ring shuffle KV shards to overlap the SDPA (Scaled Dot Product Attention) computation and the all-to-all communication
+    necessary for the next SDPA.
+
+The Context Parallel APIs consist of two parts:
+
+1.  ``context_parallel()`` allows users to create a Python context where the SDPA function (``torch.nn.functional.scaled_dot_product_attention``)
+    will be automatically replaced with Ring Attention. To shard Tensors along a dimension, simply pass the Tensors and their sharding dimensions to
+    argument ``buffers`` and ``buffer_seq_dims`` respectively. We recommend that users add tensors computing along the sequence dimension to ``buffers``
+    and shard them along this dimension.
```
Comment (fegin) [inline prototype_source/context_parallel.rst:45]: I prefer to add some examples, e.g., `For example, without adding ``freq_cis`` to ``buffers``, the rotary embedding will be miscalculated.`

**Review by fegin — APPROVED** [review]
Two final comments. LGTM

**Comment by svekars** [comment]
Can you please rebase the base branch to main.

**Review by AlannaBurke — APPROVED** [review]
